crypto/fipsmodule/sshkdf/sshkdf_test.cc

```diff
@@ -0,0 +1,522 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <gtest/gtest.h>
+
+#include <openssl/crypto.h>
+#include <openssl/digest.h>
+#include <openssl/sshkdf.h>
+
+#include "../../internal.h"
+#include "../../test/test_util.h"
+
+// SSHKDF test vectors from NIST CAVS 14.1 test vectors
+// https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Component-Testing#KDF135
+
+static void test_sshkdf(const EVP_MD *evp_md, const uint8_t *key, size_t key_len,
+    const uint8_t *xcghash, size_t xcghash_len,
+    const uint8_t *session_id, size_t session_id_len,
+    char type,
+    uint8_t *out, size_t out_len,
+    const uint8_t *expected_output) {
+    // Initial IVs
+    ASSERT_TRUE(SSHKDF(evp_md, key, key_len, xcghash, xcghash_len,
+        session_id, session_id_len, type,
+        out, out_len));
+    EXPECT_EQ(Bytes(expected_output, sizeof(out_len)),
+        Bytes(out, sizeof(out_len)));
+}
+
+TEST(SSHKDFTest, SSHKDF_TYPE_INSANITY) {
+    // The sanity check for the |type| param happens before anything else, so
+    // we can check it without any valid info.
+    ASSERT_FALSE(SSHKDF(nullptr, nullptr, 0, nullptr, 0, nullptr, 0,
```

Comment:
Don't we need to test each parameter being wrong separately?

Reply:
At the time this was written, no, but after adding more sanity checking from https://github.com/openssl/openssl/blob/ad4910fad22d57c6d685b0ca83bdb7b2bf69d8fd/providers/implementations/kdfs/sshkdf.c#L99, yes.